### PR TITLE
Fixed issue #18725: Reorder questions/question groups can break survey (Tests)

### PR DESCRIPTION
**Pull request by gabrieljenik**
- Tests Added
- A prior test was fixed
- Code Reorganization
- Added error_reporting on tests when "github logging" is enabled (this will ease diagnosis when tests fail)
- Added new flag on bootstrap for forcing error_reporting on tests

### CODE REVIEW

**Comment by Shnoulle** [comment]
I don't understand : https://github.com/LimeSurvey/LimeSurvey/pull/3092

https://bugs.limesurvey.org/view.php?id=18725 is already closed and i always keep the fix for me.

**Review by Shnoulle — COMMENTED** [review]
Already fixed

**Review by Shnoulle — CHANGES_REQUESTED** [review]
Not needed , already fixed and have test too

**Review by Shnoulle — COMMENTED** [review]
OK,
bettre understand : only add test here … let me check agian

File: tests/TestBaseClass.php [hunk 1]
```diff
@@ -37,6 +37,15 @@ public static function setUpBeforeClass(): void
     {
         parent::setUpBeforeClass();
 
+        // Enable Debug and Error Reporting if logging is enabled
```
Comment (Shnoulle) [inline tests/TestBaseClass.php:40]: Needed here or unrelated ?

File: tests/unit/helpers/FixMovedQuestionConditionsTest.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        // Import default survey
+        $filename = self::$surveysFolder . '/survey_fixMovedQuestionConditionsDefault.lss';
+        self::importSurvey($filename);
+
+        \Yii::app()->setConfig('sid', self::$surveyId);
+    }
+
+    /**
+     * Test fixing just one condition.
+     * This test uses the survey previously imported in setupBeforeClass.
```
Comment (Shnoulle) [inline tests/unit/helpers/FixMovedQuestionConditionsTest.php:22]: Maybe more detail in test ? Here : if there are an issue, i'm unsure of what is the needed result.

File: tests/unit/helpers/FixMovedQuestionConditionsTest.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                                null,
+                                null
+                            );
+
+        $firstQuestionConditions = $conditions->readAll();
+
+        $this->assertSame(
+            $firstQuestionConditions[0]['cfieldname'],
+            self::$surveyId . 'X' . $groups[0]->gid . 'X' . $questions[0]->qid,
+            'The cfieldname field is not correct.'
```
Comment (Shnoulle) [inline tests/unit/helpers/FixMovedQuestionConditionsTest.php:42]: Detail in the issue : _The cfieldname field is not correct : bad group before moving_ ?

File: tests/unit/helpers/GroupHelperTest.php [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        // Check the current order.
+        $currentOrder = \Yii::app()->db->createCommand()->select('title')
+                            ->from('{{questions}}')
+                            ->where('gid=' . $gid)
+                            ->order('question_order, title ASC')
+                            ->query()
+                            ->readAll();
+
+        $this->assertCount(3, $currentOrder, 'The number of questions in the group is not correct.');
+        $this->assertSame('Q03', $currentOrder[0]['title'], 'The question title is incorrect.');
```
Comment (Shnoulle) [inline tests/unit/helpers/GroupHelperTest.php:157]: Detail in the issue : _The question title is incorrect for 1st question after moving : have {$currentOrder[0]['title']} , wait for Q03_ ?

I don't remind if it was added automatically : waiting value, returned value. If yes : not needed to add it 2 times (but why not)./

**Review by Shnoulle — APPROVED** [review]
Seems OK, but more detail in issue in Test result can be really better.

Sometime looking at test : i unsure for what is tested and what is waiting.

File: tests/TestBaseClass.php (hunk 1 above)
Comment (gabrieljenik) [reply tests/TestBaseClass.php:40]: I was having trouble with a test and implemented this pending artifact for having more info on failing tests. I can split that into a separate PR if needed.

File: tests/bootstrap.php [hunk 5]
```diff
@@ -239,7 +239,24 @@
 $configBackupFile = __DIR__ . '/application/config/test-backup.config.php';
 
 // Enable if phpunit fails.
-// error_reporting(E_ALL);
+$forceDebug = false;
```
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply tests/bootstrap.php:242]: Same as https://github.com/LimeSurvey/LimeSurvey/pull/3178#discussion_r1211687653

**Comment by olleharstedt** [comment]
Weeeell master is supposed to be only bug fixes, and bug fixes are supposed to be as small as possible. Everything else -> dev branch.

Would rebase be trivial?

**Comment by gabrieljenik** [comment]
> Weeeell master is supposed to be only bug fixes, and bug fixes are supposed to be as small as possible. Everything else -> dev branch.
> 
> Would rebase be trivial?

Well, you are right about this not being a big fix, but it is a test related to a bug fix (and a code reorg for the test), ... 

Let you decide what to do. Shouldn't be hard to cherry pick or rebase.

File: tests/TestBaseClass.php (hunk 1 above)
Comment (Shnoulle) [reply tests/TestBaseClass.php:40]: For master : remoive it, but if in develop keep it ?

**Comment by Shnoulle** [comment]
PS : on master or develop : i still think more detail in test (comment at minima or text return).

I mean : if test broke : don't have to understand the test code, just go at the issue.

**Comment by gabrieljenik** [comment]
> Weeeell master is supposed to be only bug fixes, and bug fixes are supposed to be as small as possible. Everything else -> dev branch.
> 
> Would rebase be trivial?

What would happen to v5? We wouldn't add this test there?

**Comment by gabrieljenik** [comment]
@Shnoulle Messages updated :)

**Comment by gabrieljenik** [comment]
Rebased @olleharstedt

File: tests/unit/helpers/remotecontrol/RemoteControlQuestionPropertiesTest.php [hunk 6]
```diff
@@ -81,7 +81,7 @@ public function testGetDualQuestionProperties()
 
         $sessionKey = $this->handler->get_session_key($this->getUsername(), $this->getPassword());
 
-        $question = \Question::model()->findByAttributes(array('title' => 'G01Q02'));
+        $question = \Question::model()->findByAttributes(array('sid' => self::$surveyId,'title' => 'G01Q02'));
```
Comment (Shnoulle) [inline tests/unit/helpers/remotecontrol/RemoteControlQuestionPropertiesTest.php:84]: Yo **really needed** for local test ! :+1:

**Comment by olleharstedt** [comment]
Cool beans. :) Thanks, @gabrieljenik !

**Comment by sewalsh** [comment]
Is this getting merged to the 3.x LTS branch?

**Pull request merged**